packages/messaging/src/controllers/controller-interface.ts

```diff
@@ -74,42 +73,94 @@ export default class ControllerInterface {
       // We must wait for permission to be granted
       return Promise.resolve(null);
     }
-
-    return this.getSWRegistration_().then(registration => {
-      return this.tokenDetailsModel_
-        .getTokenDetailsFromSWScope(registration.scope)
-        .then(details => {
-          if (details) {
-            // TODO Validate the details are still accurate
-            return details['fcmToken'];
+    let swReg: ServiceWorkerRegistration;
+    return this.getSWRegistration_()
+      .then(reg => {
+        swReg = reg;
+        return this.tokenDetailsModel_.getTokenDetailsFromSWScope(swReg.scope);
+      })
+      .then(tokenDetails => {
+        if (tokenDetails) {
+          // TODO Validate the details are still accurate
+          const now = Date.now();
+          if (now < tokenDetails['createTime'] + TOKEN_EXPIRATION_MILLIS) {
+            return tokenDetails['fcmToken'];
+          } else {
+            return this.updateToken(tokenDetails, swReg);
           }
+        }
+        return this.getNewToken(swReg);
+      });
+  }
 
-          const publicVapidKey = this.getPublicVapidKey_();
-          let subscription;
-          return this.getPushSubscription_(registration, publicVapidKey)
-            .then(sub => {
-              subscription = sub;
-              return this.iidModel_.getToken(
-                this.messagingSenderId_,
-                subscription,
-                publicVapidKey
-              );
-            })
-            .then(tokenDetails => {
-              const allDetails = {
-                swScope: registration.scope,
-                vapidKey: publicVapidKey,
-                subscription: subscription,
-                fcmSenderId: this.messagingSenderId_,
-                fcmToken: tokenDetails['token'],
-                fcmPushSet: tokenDetails['pushSet']
-              };
-              return this.tokenDetailsModel_
-                .saveTokenDetails(allDetails)
-                .then(() => tokenDetails['token']);
-            });
+  private updateToken(
+    tokenDetails: Object,
+    swReg: ServiceWorkerRegistration
+  ): Promise<string> {
+    const publicVapidKey = this.getPublicVapidKey_();
+    let updatedToken: string;
+    let subscription: PushSubscription;
+    return this.getPushSubscription_(swReg, publicVapidKey)
+      .then(pushSubscription => {
+        subscription = pushSubscription;
+        return this.iidModel_.updateToken(
+          this.messagingSenderId_,
+          tokenDetails['fcmToken'],
+          tokenDetails['pushDetails'],
+          subscription,
+          publicVapidKey
+        );
+      })
+      .catch(err => {
+        return this.deleteToken(tokenDetails['fcmToken']).then(() => {
+          throw err;
         });
-    });
+      })
+      .then(token => {
+        updatedToken = token;
```

Comment:
Does this mean that every 7 days the developer will have a new token for the user? I thought this would simply check if the token should be deleted and if so it would be reset at that point, otherwise it would continue being used.

I'm just wondering if this will grow a long list of tokens in databases that are now no longer relevant.

Reply:
currently the backend is wired to always return the existing token. So technically I can just do what you have suggested. What I had in mind:
- This gives us a way from the backend to recover clients if somehow the tokens issued from FCM are problematic, ie we can roll out new non-problematic tokens.
- I was assuming that since the tokenDetails like senderId, swScope etc didn't change we'd just be overwriting the existing entry in the DB. Is that not the case?

Reply:
It'll overwrite any of the details based on the service worker scope - which won't change :)